internal/controllers/selfhealing/slice.go

```diff
@@ -0,0 +1,143 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+)
+
+const (
+	PodTimeout = time.Minute * 2
+)
+
+// sliceController check if the resource slice is deleted but it is still present in the composition current synthesis status.
+// If yes, it will update the composition PendingResynthesis status to trigger re-synthesis process.
+type sliceController struct {
+	client client.Client
+}
+
+func NewSliceController(mgr ctrl.Manager) error {
+	s := &sliceController{
+		client: mgr.GetClient(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("selfHealingSliceController").
+		Watches(&apiv1.ResourceSlice{}, newSliceHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "selfHealingSliceController")).
+		Complete(s)
+}
+
+func (s *sliceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	comp := &apiv1.Composition{}
+	err := s.client.Get(ctx, req.NamespacedName, comp)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting composition: %w", err)
+	}
+
+	syn := &apiv1.Synthesizer{}
+	syn.Name = comp.Spec.Synthesizer.Name
+	err = s.client.Get(ctx, client.ObjectKeyFromObject(syn), syn)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("gettting synthesizer: %w", err))
+	}
+
+	logger = logger.WithValues("compositionGeneration", comp.Generation,
+		"compositionName", comp.Name,
+		"compositionNamespace", comp.Namespace,
+		"synthesisID", comp.Status.GetCurrentSynthesisUUID())
+
+	// Skip if the composition is not eligible for resynthesis, and check the synthesis result later
+	if notEligibleForResynthesis(comp) {
+		logger.V(1).Info("not eligible for resynthesis when checking the missing resource slice")
+		// Use default pod timeout
+		if syn.Spec.PodTimeout == nil {
+			return ctrl.Result{Requeue: true, RequeueAfter: PodTimeout}, nil
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: syn.Spec.PodTimeout.Duration}, nil
+	}
+
+	// Check if any resource slice referenced by the composition is deleted.
+	for _, ref := range comp.Status.CurrentSynthesis.ResourceSlices {
+		slice := &apiv1.ResourceSlice{}
+		slice.Name = ref.Name
+		slice.Namespace = comp.Namespace
+		err := s.client.Get(ctx, client.ObjectKeyFromObject(slice), slice)
+		if errors.IsNotFound(err) {
+			// The resource slice should not be deleted if it is still referenced by the composition.
+			// Update the composition status to trigger re-synthesis process.
+			logger.V(1).Info("found missing resource slice and start resynthesis", "compositionName", comp.Name, "resourceSliceName", ref.Name)
+			comp.Status.PendingResynthesis = ptr.To(metav1.Now())
+			err = s.client.Status().Update(ctx, comp)
```

Comment:
Before we set `PendingResynthesis`, let's do a non-cached GET on the resource slice just to be absolutely sure it doesn't exist